### PR TITLE
Add `ImageFolderDataset`

### LINKED ISSUE #1132: [Example] Add custom images dataset
Currently we only have an example using the `MNISTDataset` via the huggingface loader.

Would be nice to include an example to load a custom image dataset from disk, similar to [`torchvision.datasets.DatasetFolder`](https://pytorch.org/vision/main/generated/torchvision.datasets.DatasetFolder.html) for image classification.

### PR DESCRIPTION
**Pull request by laggui**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.

### Related Issues/PRs

Closes #1132

### Changes

- Added `ImageFolderDataset` to load images from disk (following a folder structure)
- Added a `custom-image-dataset` example that downloads CIFAR-10 and trains a simple CNN model using the `ImageFolderdataset`

### Testing

Added a couple of unit tests and an example.

### CODE REVIEW

**Comment by laggui** [comment]
Note that I added an `ImageTarget` enum for image classification, object detection and segmentation targets. At this time only the image classification targets are implemented while the other two would be for the future.

We could remove the options until they are implemented. Let me know your thoughts :)

File: burn-dataset/src/vision/image_folder.rs [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+pub struct SegmentationMask {
+    /// Segmentation mask.
+    pub mask: Vec<usize>,
+}
+
+/// Object detection bounding box target.
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
+pub struct BoundingBox {
+    /// Coordinates.
+    pub coords: [f32; 4],
```
Comment (nathanielsimard) [inline burn-dataset/src/vision/image_folder.rs:83]: Are the coordinates `f32` or are they supposed to be int? Not sure what is the most common.

File: burn-dataset/src/vision/image_folder.rs [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+struct ImageDatasetItemRaw {
+    /// Image path.
+    pub image_path: PathBuf,
+
+    /// Image target.
+    /// The target string can be a category name or path to annotation file.
+    pub target: String,
```
Comment (nathanielsimard) [inline burn-dataset/src/vision/image_folder.rs:106]: If the target can be parsed into all `ImageTarget` I would avoid using a string, but consider it as a byte array `Vec<u8>`, so that it is explicit that we need to decode it. If it's only used for classification, I would change the type to `ImageClassificationDatasetItemRaw`.

File: burn-dataset/src/vision/image_folder.rs [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+        self.dataset.get(index)
+    }
+
+    fn len(&self) -> usize {
+        self.dataset.len()
+    }
+}
+
+impl ImageFolderDataset {
+    /// Create an image classification dataset from the root folder.
```
Comment (nathanielsimard) [inline burn-dataset/src/vision/image_folder.rs:206]: Since the dataset type can be used for other things than classification, should we rename the doc or the method? If the method is only valid for classification, I would rename to `new_classification`, otherwise just update the doc.

File: burn-dataset/src/vision/image_folder.rs [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    /// * `root` - Dataset root folder.
+    ///
+    /// # Returns
+    /// A new dataset instance.
+    pub fn new<P: AsRef<Path>>(root: P) -> Self {
+        // New dataset containing any of the supported file types
+        ImageFolderDataset::new_with(root, &SUPPORTED_FILES)
+    }
+
+    /// Create an image classification dataset from the root folder.
```
Comment (nathanielsimard) [inline burn-dataset/src/vision/image_folder.rs:219]: Same comment as with `new`

File: burn-dataset/src/vision/image_folder.rs (hunk 2 above)
Comment (nathanielsimard) [reply burn-dataset/src/vision/image_folder.rs:106]: Also, for classification, we may have an enum for the classification target, since a lot of time it is the label, but other time it can be the index. For the raw part we should support both ways, but when preprocessed, I think we should only expose the index.

File: burn-dataset/src/vision/image_folder.rs [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+                    .map(check_extension)
+                    .collect::<Vec<_>>()
+                    .join(",")
+            )],
+        )
+        .follow_links(true)
+        .sort_by(|p1: &DirEntry, p2: &DirEntry| p1.path().cmp(p2.path())) // order by path
+        .build()
+        .unwrap()
+        .filter_map(Result::ok);
```
Comment (nathanielsimard) [inline burn-dataset/src/vision/image_folder.rs:260]: I believe we should return a `Result` from the dataset creation. Similar to the Hugging Face dataset loader, certain errors may occur that are not programmer errors, such as file disk access issues. Therefore, implementing error handling in this section is a good practice.

File: examples/custom-image-dataset/src/data.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                    .swap_dims(1, 0) // [C, H, W]
+            })
+            .map(|tensor| tensor / 255) // normalize between [0, 1]
+            .collect();
+
+        let images = Tensor::stack(images, 0);
+        let targets = Tensor::cat(targets, 0);
+
+        let normalizer = Normalizer::<B>::new(&self.device);
+        let images = normalizer.normalize(images);
```
Comment (nathanielsimard) [inline examples/custom-image-dataset/src/data.rs:93]: The normalizer should be included in the batcher as a field I think.

```rust
pub struct ClassificationBatcher<B: Backend> {
    device: B::Device,
    normalizer: Normalizer<B>,
}
```

File: examples/custom-image-dataset/src/downloader.rs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+use reqwest::Client;
+use std::cmp::min;
+use std::io::Write;
+
+/// Download the file at the specified url to a bytes vector.
+/// File download progress is reported with the help of a [progress bar](indicatif).
+///
+/// Taken from [burn-dataset](https://github.com/tracel-ai/burn/blob/main/burn-dataset/src/vision/downloader.rs).
+#[tokio::main(flavor = "current_thread")]
+pub async fn download_file_as_bytes(url: &str, message: &str) -> Vec<u8> {
```
Comment (nathanielsimard) [inline examples/custom-image-dataset/src/downloader.rs:11]: It seems to be duplicated from the `mnist` dataset implementation right? Maybe we can actually add cifar to `burn-dataset` and reuse the setup from mnist?

File: examples/custom-image-dataset/examples/custom-image-dataset.rs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        let device = LibTorchDevice::Cuda(0);
+        #[cfg(target_os = "macos")]
+        let device = LibTorchDevice::Mps;
+
+        train::<Autodiff<LibTorch>>(
+            TrainingConfig::new(SgdConfig::new().with_momentum(Some(MomentumConfig {
+                momentum: 0.9,
+                dampening: 0.,
+                nesterov: false,
+            }))),
```
Comment (nathanielsimard) [inline examples/custom-image-dataset/examples/custom-image-dataset.rs:20]: The question is: is Adam better in this case 😅

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Very cool and complete. I have a few comments to access before merging, but great job! 👏

File: burn-dataset/src/vision/image_folder.rs (hunk 2 above)
Comment (laggui) [reply burn-dataset/src/vision/image_folder.rs:106]: Agreed, this was supposed to target only classification. Will rename to `ImageClassificationDatasetItemRaw`. Also totally forgot about the index which could be provided instead of the string.

File: burn-dataset/src/vision/image_folder.rs (hunk 3 above)
Comment (laggui) [reply burn-dataset/src/vision/image_folder.rs:206]: Yes! Will rename to `new_classification` since we only implement loading classification labels at this time.

File: burn-dataset/src/vision/image_folder.rs (hunk 5 above)
Comment (laggui) [reply burn-dataset/src/vision/image_folder.rs:260]: That's the type of feedback I was looking for 🙏 I am still not always sure when it is better practice to use `unwrap()` or other such methods vs returning a `Result`.

File: examples/custom-image-dataset/src/downloader.rs (hunk 7 above)
Comment (laggui) [reply examples/custom-image-dataset/src/downloader.rs:11]: Well the code is technically not duplicated since both call that function which is separated in its own module :) but yes, both use the same function (intentionally) and that download function can be used for any other future dataset we might want to add.

I wasn't sure if we wanted to add CIFAR-10 explicitly so I only put it as an example, but it could be refactored. And actually the original CIFAR-10 source is in bytes form (I used a mirror with the folder structure to illustrate the `ImageFolderDataset` which is more common).

File: examples/custom-image-dataset/examples/custom-image-dataset.rs (hunk 8 above)
Comment (laggui) [reply examples/custom-image-dataset/examples/custom-image-dataset.rs:20]: I actually haven't fiddled with the training recipe too much for this example 😄 a bunch of things could probably be changed to further optimize the results, I just wanted to keep it simple.

Lmk if you want any changes to be applied to the training example.

File: burn-dataset/src/vision/image_folder.rs (hunk 2 above)
Comment (laggui) [reply burn-dataset/src/vision/image_folder.rs:106]: Also, regarding `ImageTarget` I just realized it doesn't generalize very well for object detection. I think I will rename it to `Annotation` instead.

File: burn-dataset/src/vision/image_folder.rs (hunk 5 above)
Comment (nathanielsimard) [reply burn-dataset/src/vision/image_folder.rs:260]: `unwrap` => dev errors (made by us or the user such as dimension mismatch in the tensor API) 

`runtime error` => Result

**Comment by laggui** [comment]
Thanks for the feedback! Wasn't clear if you wanted to add CIFAR-10 to `burn-dataset` immediately or not. If that is the case just say so and I'll refactor it.

**Review by nathanielsimard — APPROVED** [review]
LGTM, we could add CIFAR to the burn-dataset crates, but since you mentioned you used the folder version as an example, maybe it's best to just keep it in the example dir. Maybe we can add another version more similar to MNIST in the crate.
